language-bindings/rust/samples/wasi-hello/src/main.rs

```diff
@@ -0,0 +1,128 @@
+        0x10, 0x00, 0x1A, 0x20, 0x01, 0x41, 0x04, 0x6A, 0x41, 0x8E, 0x09, 0x2F, 0x00, 0x00, 0x3B,
+        0x00, 0x00, 0x20, 0x01, 0x41, 0x8A, 0x09, 0x28, 0x00, 0x00, 0x36, 0x00, 0x00, 0x20, 0x00,
+        0x20, 0x01, 0x36, 0x02, 0x00, 0x41, 0x80, 0x08, 0x20, 0x00, 0x10, 0x00, 0x1A, 0x20, 0x01,
+        0x10, 0x03, 0x41, 0x00, 0x0B, 0x20, 0x00, 0x41, 0x20, 0x6A, 0x24, 0x00, 0x0B, 0x0B, 0x97,
+        0x01, 0x01, 0x00, 0x41, 0x80, 0x08, 0x0B, 0x8F, 0x01, 0x62, 0x75, 0x66, 0x3A, 0x20, 0x25,
+        0x73, 0x00, 0x6D, 0x61, 0x6C, 0x6C, 0x6F, 0x63, 0x20, 0x62, 0x75, 0x66, 0x20, 0x66, 0x61,
+        0x69, 0x6C, 0x65, 0x64, 0x00, 0x48, 0x65, 0x6C, 0x6C, 0x6F, 0x20, 0x77, 0x6F, 0x72, 0x6C,
+        0x64, 0x21, 0x00, 0x62, 0x75, 0x66, 0x20, 0x70, 0x74, 0x72, 0x3A, 0x20, 0x25, 0x70, 0x0A,
+        0x00, 0x25, 0x66, 0x20, 0x2A, 0x20, 0x25, 0x66, 0x20, 0x3D, 0x20, 0x25, 0x66, 0x0A, 0x00,
+        0x25, 0x66, 0x20, 0x2F, 0x20, 0x25, 0x64, 0x20, 0x3D, 0x20, 0x25, 0x66, 0x0A, 0x00, 0x69,
+        0x33, 0x32, 0x3A, 0x20, 0x25, 0x64, 0x2C, 0x20, 0x69, 0x36, 0x34, 0x3A, 0x20, 0x25, 0x6C,
+        0x6C, 0x64, 0x2C, 0x20, 0x66, 0x33, 0x32, 0x3A, 0x20, 0x25, 0x66, 0x2C, 0x20, 0x66, 0x36,
+        0x34, 0x3A, 0x20, 0x25, 0x66, 0x0A, 0x00, 0x25, 0x6C, 0x6C, 0x64, 0x20, 0x2B, 0x20, 0x25,
+        0x6C, 0x6C, 0x64, 0x20, 0x3D, 0x20, 0x25, 0x6C, 0x6C, 0x64, 0x0A, 0x00, 0x31, 0x32, 0x33,
+        0x34, 0x0A,
+    ];
+
+    let payload_ptr = payload.clone().as_mut_ptr();
+
+    let mut error_buf = [0i8; DEFAULT_ERROR_BUF_SIZE];
+
+    let ret = unsafe { wasm_runtime_init() };
+    assert!(ret);
+
+    let module = unsafe {
+        wasm_runtime_load(
+            payload_ptr,
+            payload.len() as u32,
+            error_buf.as_mut_ptr(),
+            error_buf.len() as u32,
+        )
+    };
+
+    assert!((module as usize) != 0);
+
+    unsafe {
+        wasm_runtime_set_wasi_args(
+            module,
+            std::ptr::null_mut(),
+            0,
+            std::ptr::null_mut(),
+            0,
+            std::ptr::null_mut(),
+            0,
+            std::ptr::null_mut(),
+            0,
+        );
+    }
+
+    let module_instance = unsafe {
+        wasm_runtime_instantiate(
+            module,
+            DEFAULT_STACK_SIZE,
+            DEFAULT_HEAP_SIZE,
+            error_buf.as_mut_ptr(),
+            error_buf.len() as u32,
+        )
+    };
+
+    //let err_u8_vec: Vec<u8> = error_buf.iter().map(|&x| x as u8).filter(|x| *x > 31 && *x < 124).collect();
+    //print!("error {:?}", String::from_utf8(err_u8_vec));
+
+    assert!((module_instance as usize) != 0);
+
+    let _exec_env = unsafe { wasm_runtime_create_exec_env(module_instance, DEFAULT_STACK_SIZE) };
+
+    let success =
+        unsafe { wasm_application_execute_main(module_instance, 0, std::ptr::null_mut()) };
+
+    assert!(success);
+
+    let _main_result = unsafe { wasm_runtime_get_wasi_exit_code(module_instance) };
```

Comment:
Not sure whether we should wasm_runtime_deinstantiate(module_inst), wasm_runtime_unload(module) and wasm_runtime_destroy() before return in rust?
See
https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/product-mini/platforms/posix/main.c#L978-L999

Reply:
You're right, even if this is just an example, it should clean up in the end.